rfcs/79-minimal-fragmenting-relationships.md

```diff
@@ -0,0 +1,128 @@
+# Feature Name: `minimal-fragmenting-relationships`
+
+## Summary
+
+The central idea behind fragmenting relationships is the ability to express a connection between two `Entity`s by adding a pair of `(ComponentId, Entity)` as if it were a component. For example if one had several entities representing different kinds of swords one could add the pair `(Equipped, sword_entity)` to the player to signify which is currently equipped. 
+
+Of course you could also just have created an `Equipped` component with an `Entity` field member to represent the same relationship. The advantage to "fragmenting" relationships is each unique pair is treated as if it were a unique component id, therefore they can appear in queries, use the existing insert and get APIs as well as form thier own archetypes. For example we could add `(Equipped, shield_entity)` to the same player and query for `(Equipped, *)`to find all equipped pairs.
+
+Another analogy when considering the ECS as an in memory database is that relationships act as foreign keys. This unlocks a potential future where we can express more complex queries matching across multiple entities and relationship pairs. There are a myriad of additional features that can be built on top of the initial implementation described below. The main focus in this RFC will be the blocking issues that must be addressed before the initial implementation is achievable.
+## Motivation
+
+Relationships have been a long desired bevy feature. The ability to express and traverse hierarchies and graphs within the ECS unlocks a huge amount of potential. Features such as breadth-first query traversal, immediate hierarchy clean-up, component inheritance, multi-target queries etc. all rely on or are simplified by the existence of relationships as a first class ECS primitive. 
+## User-facing explanation
+
+Relationship pairs can be inserted and removed like any other component. There are 3 ids associated with any relationship pair instance:
+1. The source entity that the pair is being added to
+2. The relationship component
+3. The target entity
+
+In the minimal implementation all relationships will take the form of `(ZST Component, entity)` pairs so the API surface will look quite familiar. All naming is subject to bikeshedding:
+```rust
+#[derive(Component)]
+struct Relationship;
+
+world.entity_mut(source_entity_a)
+	.insert_pair::<Relationship>(target_entity_b)
```

Comment:
I'm not keen on the naming here. A "pair" is a new concept / terminology that I don't think actually adds any expressive value over `insert_relation`.

Reply:
Very fair, this is me bringing in baggage from other languages that defined two item tuples as pairs. I think the main reason I went that way over `insert_relation` is just how long that is for a very common operation that already needs a type parameter.

Reply:
Fwiw the reason I picked pair is because it's less confusing/opinionated:
- Does relation(ship) refer to the first element of the pair or both the relationship and the target?
- The feature is called relationships, but it really implements the ability to have composite component ids which can be used for relationships. I can use it to add a `(Dog, Food)` pair, where `Dog` isn't really a relationship and `Food` isn't really a relationship target.

Reply:
I can get behind that. The key thing here is clearly laying out (and motivating) the terminology.